tortoise/verifying_tortoise.go

```diff
@@ -130,492 +242,1007 @@ func blockIDsToString(input []types.BlockID) string {
-			if _, ok := inInputVector[b]; ok {
-				continue
-			}
-
-			// TODO: maybe we don't need this if it is not included
-			if v, ok := opinion2.BlocksOpinion[b]; !ok || v != against {
-				// Layer block has no base block vote or base block supports, but not in input vector:
-				// add diff AGAINST this block
-				t.logger.With().Debug("added diff",
-					log.FieldNamed("base_block_candidate", blockid),
-					log.FieldNamed("diff_block", b),
-					log.String("diff", "against"))
-				againstDiff[b] = struct{}{}
-			}
-		}
+		// Finally, we need to consider the case where the base block supports a block in this layer that is not in our
+		// input vector (e.g., one we haven't seen), by adding a diff against the block.
+		// We do not explicitly add votes against blocks that the base block does _not_ support, since by not voting to
+		// support a block in an old layer, we are implicitly voting against it. But if the base block does explicitly
+		// support a block and we disagree, we need to add a vote against here.
+		// TODO: this is not currently possible since base block opinions aren't indexed by layer. See
+		//   https://github.com/spacemeshos/go-spacemesh/issues/2424
+		//for b, v := range baseBlockOpinion.BlockOpinions {
+		//	if _, ok := inInputVector[b]; !ok && v != against {
+		//		addDiffs(b, "against", against, againstDiff)
+		//	}
+		//}
 	}
 
 	// check if exceeded max no. exceptions
 	explen := len(againstDiff) + len(forDiff) + len(neutralDiff)
 	if explen > t.MaxExceptions {
-		return nil, fmt.Errorf("too many exceptions to base block vote (%v)", explen)
+		return nil, fmt.Errorf("%s (%v)", errstrTooManyExceptions, explen)
 	}
 
 	return []map[types.BlockID]struct{}{againstDiff, forDiff, neutralDiff}, nil
 }
 
-func (t *turtle) BlockWeight(voting, voted types.BlockID) int {
-	return 1
+// voteWeight returns the weight to assign to one block's vote for another.
+// Note: weight depends on more than just the weight of the voting block. It also depends on contextual factors such as
+// whether or not the block's ATX was received on time, and on how old the layer is.
+// TODO: for now it's probably sufficient to adjust weight based on whether the ATX was received on time, or late, for
+//   the current epoch. See https://github.com/spacemeshos/go-spacemesh/issues/2540.
+func (t *turtle) voteWeight(ctx context.Context, votingBlock *types.Block) (uint64, error) {
+	logger := t.logger.WithContext(ctx)
+
+	atxHeader, err := t.atxdb.GetAtxHeader(votingBlock.ATXID)
+	if err != nil {
+		return 0, err
+	}
+	atxTimestamp, err := t.atxdb.GetAtxTimestamp(votingBlock.ATXID)
+	if err != nil {
+		return 0, err
+	}
+	atxEpoch := atxHeader.PubLayerID.GetEpoch()
+	nextEpochStart := t.clock.LayerToTime((atxEpoch + 1).FirstLayer())
+
+	// check if the ATX was received on time
+	// TODO: add an exception for sync, when we expect everything to be received late
+	if atxTimestamp.Before(nextEpochStart) {
```

Comment:
@lrettig this appears to be the reason why the verified layer was stuck on my new devnet. it might be as well the reason for latenodes failure, or i misunderstood something?

Reply:
good point, I'm testing without this